pex/pip/local_project.py

```diff
@@ -0,0 +1,62 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import
+
+import os.path
+import tarfile
+
+from pex import hashing
+from pex.build_system import pep_517
+from pex.common import temporary_dir
+from pex.hashing import Fingerprint, Sha256
+from pex.resolve.resolvers import Resolver
+from pex.result import Error, try_
+from pex.tracer import TRACER
+from pex.typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from typing import Optional, Union
+
+    from pex.hashing import HintedDigest
+
+
+def fingerprint_local_project(
+    directory,  # type: str
+    resolver,  # type: Resolver
+):
+    # type: (...) -> Fingerprint
+    digest = Sha256()
+    try_(digest_local_project(directory, digest, resolver))
+    return digest.hexdigest()
+
+
+def digest_local_project(
+    directory,  # type: str
+    digest,  # type: HintedDigest
+    resolver,  # type: Resolver
+    dest_dir=None,  # type: Optional[str]
+):
+    # type: (...) -> Union[str, Error]
+    with TRACER.timed("Fingerprinting local project at {directory}".format(directory=directory)):
+        with temporary_dir() as td:
+            sdist_or_error = pep_517.build_sdist(
+                project_directory=directory,
+                dist_dir=os.path.join(td, "dists"),
+                resolver=resolver,
+            )
+            if isinstance(sdist_or_error, Error):
+                return sdist_or_error
+            sdist = sdist_or_error
+
+            extract_dir = dest_dir or os.path.join(td, "extracted")
```

Comment:
When `dest_dir` is set, it means it's possible for this function to be called multiple times in parallel and to write to the same dest. What's the concurrency safety of this?

Reply:
As you spotted, it's unsafe generally as an API but its safe as used.

There are 2 callers:
1. `fingerprint_local_project` above which does not pass a `dest`
2. `LocalProjectDownloadManager.save` which is given an exclusively locked atomic_directory by `DownloadManager.store` (downloads happen exactly-once).

Reply:
I'm going to merge, but I'm also not happy about this non-locally safe property. I'm thinking through ways to tighten this up so a local read shows this is always safe.

Reply:
Thanks for explaining. In the absence of a more elegant design...something like requiring the user to set `called_under_lock=True` if `dest_dir` is set could work. Like Rust's `unsafe!` declaration.